crates/engine/tree/src/tree/root.rs

```diff
@@ -673,6 +673,47 @@ where
         }
     }
 
+    /// Starts the main loop that handles all incoming messages, fetches proofs, applies them to the
+    /// sparse trie, updates the sparse trie, and eventually returns the state root.
+    ///
+    /// The lifecycle is the following:
+    /// 1. Either [`StateRootMessage::PrefetchProofs`] or [`StateRootMessage::StateUpdate`] is
+    ///    received from the engine.
+    ///    * For [`StateRootMessage::StateUpdate`], the state update is hashed with
+    ///      [`evm_state_to_hashed_post_state`], and then (proof targets)[`MultiProofTargets`] are
+    ///      extracted with [`get_proof_targets`].
+    ///    * For both messages, proof targets are deduplicated according to `fetched_proof_targets`,
+    ///      so that the proofs for accounts and storage slots that were already fetched are not
+    ///      requested again.
+    /// 2. Using the proof targets, a new multiproof is calculated using
+    ///    [`MultiproofManager::spawn_or_queue`].
+    ///    * If the list of proof targets is empty, the [`StateRootMessage::EmptyProof`] message is
+    ///      sent back to this task along with the original state update.
+    ///    * Otherwise, the multiproof is calculated and the [`StateRootMessage::ProofCalculated`]
+    ///      message is sent back to this task along with the resulting multiproof, proof targets
+    ///      and original state update.
+    /// 3. Either [`StateRootMessage::EmptyProof`] or [`StateRootMessage::ProofCalculated`] is
+    ///    received.
+    ///    * The multiproof is added to the (proof sequencer)[`ProofSequencer`].
+    ///    * If the proof sequencer has a contiguous sequence of multiproofs in the same order as
+    ///      state updates arrived (i.e. transaction order), such sequence is returned.
```

Comment:
`such sequence is returned.` - can you elaborate on this? Does this basically mean we're adding proofs to the sequencer until we have a contiguous sequence and only then we go to step 4?